lib/rex/proto/dns/packet.rb

```diff
@@ -0,0 +1,278 @@
+  #
+  # @return [Resolv::DNS::Message]
+  def self.encode_raw(packet)
+    return packet unless packet.respond_to?(:decode) or packet.respond_to?(:data)
+    packet.respond_to?(:data) ? packet.data : packet.encode
+  end
+
+  #
+  # Generates a request packet, taken from Net::DNS::Resolver
+  #
+  # @param subject [String] Subject name of question section
+  # @param type [Fixnum] Type of DNS record to query
+  # @param cls [Fixnum] Class of dns record to query
+  # @param recurse [Fixnum] Recursive query or not
+  #
+  # @return [Net::DNS::Packet] request packet
+  def self.generate_request(subject, type = Net::DNS::A, cls = Net::DNS::IN, recurse = 1)
+    case subject
+    when IPAddr
+      name = subject.reverse
+      type = Net::DNS::PTR
+    when /\d/ # Contains a number, try to see if it's an IP or IPv6 address
+      begin
+        name = IPAddr.new(subject).reverse
+        type = Net::DNS::PTR
+      rescue ArgumentError
+        name = subject if self.valid_hostname?(subject)
+      end
+    else
+      name = subject if self.valid_hostname?(subject)
+    end
+
+    # Create the packet
+    packet = Net::DNS::Packet.new(name,type,cls)
+
+    if packet.query?
+      packet.header.recursive = recurse
+    end
+
+    # DNSSEC and TSIG stuff to be inserted here
+
+    return packet
+  end
+
+  #
+  # Generates a response packet for an existing request
+  #
+  # @param request [String] Net::DNS::Packet, Resolv::DNS::Message] Original request
+  # @param answer [Array] Set of answers to provide in the response
+  #
+  # @return [Net::DNS::Packet] Response packet
+  def self.generate_response(request, answer = nil)
+    packet = self.encode_net(request)
+    packet.answer = answer unless answer.nil?
+    # Set answer count header section
+    packet.header.anCount = packet.answer.count
+    # Set error code for NXDomain or unset it if reprocessing a response
+    if packet.header.anCount < 1
+      packet.header.rCode = 3
+    else
+      if packet.header.response? and packet.header.rCode.code == 3
+        packet.header.rCode = 0
+      end
+    end
+    # Set response bit last to allow reprocessing of responses
+    packet.header.qr = 1
+    return packet
+  end
+
+  module Raw
+
+    #
+    # Convert data to little endian unsigned short
+    #
+    # @param data [Fixnum, Float, Array] Input for conversion
+    #
+    # @return [String] Raw output
+    def self.to_short_le(data)
+      [data].flatten.pack('S*')
+    end
+
+    #
+    # Convert data from little endian unsigned short
+    #
+    # @param data [String] Input for conversion
+    #
+    # @return [Array] Integer array output
+    def self.from_short_le(data)
+      data.unpack('S*')
+    end
+
+    #
+    # Convert data to little endian unsigned int
+    #
+    # @param data [Fixnum, Float, Array] Input for conversion
+    #
+    # @return [String] Raw output
+    def self.to_int_le(data)
+      [data].flatten.pack('I*')
```

Comment:
All of these host-order pack operations are wrong on big-endian platforms (need v and V, not sS and iI, or to use the byte order specifier).